brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -6,81 +6,145 @@ import com.twitter.algebird._
+        }
+
+        r
+      }
+    }
+  }
+
+  def containsKey(key: A): Boolean = randValue(key) <= threshold
+  def update(key: A, value: B) {
+    if(containsKey(key))
+      mapValues += key -> value
+  }
+
+  def get(key: A): Option[B] = mapValues.get(key)
+}
+
 case class Trainer[K: Ordering, V: Ordering, T: Monoid](
     trainingData: Iterable[Instance[K, V, T]],
     sampler: Sampler[K],
     trees: List[Tree[K, V, T]])(implicit traversal: AnnotatedTreeTraversal[K, V, T, Unit]) {
 
-  private def updateTrees(fn: (Tree[K, V, T], Int, Map[(Int, T, Unit), Iterable[Instance[K, V, T]]]) => Tree[K, V, T]): Trainer[K, V, T] = {
-    val newTrees = trees.zipWithIndex.par.map {
-      case (tree, index) =>
-        val byLeaf =
-          trainingData.flatMap { instance =>
-            val repeats = sampler.timesInTrainingSet(instance.id, instance.timestamp, index)
-            if (repeats > 0) {
-              tree.leafFor(instance.features).map { leaf =>
-                (1 to repeats).toList.map { i => (instance, leaf) }
-              }.getOrElse(Nil)
-            } else {
-              Nil
-            }
-          }.groupBy { _._2 }
-            .mapValues { _.map { _._1 } }
-        fn(tree, index, byLeaf)
+  val treeMap = trees.zipWithIndex.map{case (t,i) => i->t}.toMap
+
+  def expand(maxLeavesPerTree: Int)(implicit splitter: Splitter[V, T], evaluator: Evaluator[V, T], stopper: Stopper[T]): Trainer[K, V, T] = {
+    val allStats = treeMap.map{case (treeIndex, tree) =>
+      treeIndex -> SampledMap[Int,Map[K,splitter.S]](maxLeavesPerTree)
     }
-    copy(trees = newTrees.toList)
-  }
 
-  private def updateLeaves(fn: (Int, (Int, T, Unit), Iterable[Instance[K, V, T]]) => Node[K, V, T, Unit]): Trainer[K, V, T] = {
-    updateTrees {
-      case (tree, treeIndex, byLeaf) =>
-        val newNodes = byLeaf.map {
-          case (leaf, instances) =>
-            val (index, _, _) = leaf
-            index -> fn(treeIndex, leaf, instances)
+    trainingData.foreach{instance =>
+      lazy val features = instance.features.mapValues { value => splitter.create(value, instance.target) }
+
+      for (
+        (treeIndex, tree) <- treeMap.toList;
+        treeStats <- allStats.get(treeIndex).toList;
+        i <- 1.to(sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)).toList;
+        (leafIndex, target, annotation) <- tree.leafFor(instance.features).toList
+          if stopper.shouldSplit(target) && treeStats.containsKey(leafIndex);
+        (feature, stats) <- features
+          if sampler.includeFeature(feature, treeIndex, leafIndex)
+      ) {
+        var leafStats = treeStats.get(leafIndex).getOrElse(Map[K,splitter.S]())
+        val combined = leafStats.get(feature) match {
+          case Some(old) => splitter.semigroup.plus(old, stats)
+          case None => stats
         }
+        leafStats += feature -> combined
+        treeStats.update(leafIndex, leafStats)
+      }
+    }
```

Comment:
Since you are using this imperatively, I think it would be more efficient to do this imperatively (you wouldn't need to call `.toList` internally everywhere):

```scala
treeMap.foreach { case (treeIndex, tree) =>
  allStats.get(treeIndex).foreach { treeStats =>
    val times = sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)
    (1 to times).foreach { i =>
      tree.leafFor(instance.features).foreach { case (leafIndex, target, annotation) =>
        if (stopper.shouldSplit(target) && treeStats.containsKey(leafIndex)) {
          features.foreach { case (feature, stats) =>
            if sampler.includeFeature(feature, treeIndex, leafIndex) {
              var leafStats = treeStats.get(leafIndex).getOrElse(Map[K,splitter.S]())
              val combined = leafStats.get(feature) match {
                case Some(old) => splitter.semigroup.plus(old, stats)
                case None => stats
              }
              leafStats += feature -> combined
              treeStats.update(leafIndex, leafStats)
            }
          }
        }
      }
    }
  }
}
```